.github/workflows/full_stack_tests.yml

```diff
@@ -189,105 +189,136 @@ jobs:
           VIDEO_RECORDING_IS_ENABLED=1
           xvfb-run -a --server-args="-screen 0, 1920x1080x24"
           python -m scripts.run_acceptance_tests
           --skip-build --suite=${{ matrix.suite.name }} --prod_env --server_log_level=info
           | tee desktop_acceptance_test_retry_${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}.log
+
       - name: Upload desktop acceptance retry test logs
-        if: ${{ failure() }}
+        if: ${{ steps.desktop-test-attempt-2.outcome == 'failure' }}
         uses: actions/upload-artifact@v4
         with:
           name: desktop-acceptance-test-retry-logs-${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}
           path: desktop_acceptance_test_retry_${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}.log
-      - name: Uploading Desktop test recordings as Artifacts
-        if: ${{ failure() }}
+      
+      - name: Upload desktop test recordings as an artifact
+        # We are checking for failure of desktop-test-attempt-1 and not desktop-test-attempt-2 because
+        # because we want to upload video recording regardless of whether desktop-test-attempt-2 passes or fails.
+        if: ${{ steps.desktop-test-attempt-1.outcome == 'failure' }}
         uses: actions/upload-artifact@v4
         with:
           name: desktop-acceptance-test-video-${{steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME}}
           path: /home/runner/work/oppia/oppia_full_stack_test_video_recordings/acceptance
-      - name: Uploading generated test to angular module mapping as an artifact
-        if: ${{ failure() }}
+
+        # TODO(#22667): We are running this step twice, as snapshot gets deleted if 2nd attempt passes.
+      - name: Upload diff screenshots as an artifact
+        if: ${{ steps.desktop-test-attempt-2.outcome == 'failure' }}
         uses: actions/upload-artifact@v4
         with:
-          name: generated-test-to-angular-module-mapping-${{steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME}}
-          path: /home/runner/work/oppia/oppia/core/tests/test-modules-mappings/acceptance/${{ steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME }}.txt
+          name: diff-snapshots-desktop-attempt-2-${{steps.generate_suite_name_for_artifacts.outputs.MODIFIED_SUITE_NAME}}
+          # Note: The path must be kept in sync with the one specified in
+          # core/tests/puppeteer-acceptance-tests/utilities/common/puppeteer-utils.ts
+          path: /home/runner/work/oppia/oppia/core/tests/puppeteer-acceptance-tests/diff-snapshots
+
+      # Mobile Test (Attempt 1)
       - name: Run Mobile Acceptance Test ${{ matrix.suite.name }}
+        id: mobile-test-attempt-1
         # TODO(#22448): Remove once all suites are fixed.
-        if: ${{ steps.check_skip.outputs.SKIP_SUITE != 'true' }}
+        if: ${{ steps.check_skip.outputs.SKIP_SUITE != 'true' && always() }}
```

Comment:
Why do we need to add always()?

Reply:
If we don't add always, it will be skipped if Desktop test fails.

Reply:
Ah ok, good catch. Maybe add a comment to that effect since it's probably not obvious.

Reply:
Done.